File: DailyCore/Sources/DailyCore/Models/UserProfile.swift
import Foundation

public enum AuthProvider: String, Codable, Sendable {
    case google = "google"
    case apple = "apple"
    case email = "email"
    case guest = "guest"
    
    public var displayName: String {
        switch self {
        case .google: return "Google"
        case .apple: return "Apple"
        case .email: return "Email"
        case .guest: return "Guest"
        }
    }
}

public struct UserProfile: Codable, Equatable, Sendable, Identifiable {
    public let id: String
    public let email: String?
    public let fullName: String?
    public let avatarUrl: String?
    public let provider: AuthProvider
    public let createdAt: Date?
    
    public init(
        id: String,
        email: String? = nil,
        fullName: String? = nil,
        avatarUrl: String? = nil,
        provider: AuthProvider = .guest,
        createdAt: Date? = nil
    ) {
        self.id = id
        self.email = email
        self.fullName = fullName
        self.avatarUrl = avatarUrl
        self.provider = provider
        self.createdAt = createdAt
    }
    
    /// Derives the user's first name using the same heuristic as WinUI's `WinUIAuthService.CurrentUserFirstName`.
    public var firstName: String {
        if let fullName = fullName?.trimmingCharacters(in: .whitespacesAndNewlines), !fullName.isEmpty {
            let parts = fullName.split(separator: " ")
            if let first = parts.first, !first.isEmpty {
                return String(first)
            }
        }
        
        if let email = email, !email.isEmpty {
            let localPart = email.split(separator: "@").first ?? ""
            let nameParts = localPart.split { $0 == "." || $0 == "_" || $0 == "-" }
            if let first = nameParts.first, !first.isEmpty {
                return first.prefix(1).uppercased() + first.dropFirst()
            }
        }
        
        return provider == .guest ? "Friend" : "User"
    }
    
    public static let guest = UserProfile(
        id: "guest_\(UUID().uuidString.prefix(8))",
        email: nil,
        fullName: "Guest User",
        avatarUrl: nil,
        provider: .guest,
        createdAt: Date()
    )
}

public enum AuthSessionState: Equatable, Sendable {
    case initializing
    case unauthenticated
    case authenticated(UserProfile)
    case guest
    
    public var isAuthenticatedOrGuest: Bool {
        switch self {
        case .authenticated, .guest:
            return true
        default:
            return false
        }
    }
    
    public var profile: UserProfile? {
        switch self {
        case .authenticated(let user):
            return user
        case .guest:
            return .guest
        default:
            return nil
        }
    }
}

File: Android/core-model/src/main/java/com/intellidream/daily/model/UserProfile.kt
package com.intellidream.daily.model

import kotlinx.serialization.Serializable

@Serializable
enum class AuthProvider(val value: String, val displayName: String) {
    Google("google", "Google"),
    Apple("apple", "Apple"),
    Email("email", "Email"),
    Guest("guest", "Guest")
}

@Serializable
data class UserProfile(
    val id: String,
    val email: String? = null,
    val fullName: String? = null,
    val avatarUrl: String? = null,
    val provider: AuthProvider = AuthProvider.Guest,
    val createdAt: Long? = null
) {
    /**
     * Derives user's first name using the exact heuristic from WinUI & iOS:
     * 1. Splits fullName by spaces, picks first segment.
     * 2. If absent, parses email local part (splitting on '.', '_', '-') and capitalizes leading token.
     * 3. Fallback: "User" (or "Friend" if guest).
     */
    val firstName: String
        get() {
            val trimmedName = fullName?.trim()
            if (!trimmedName.isNullOrEmpty()) {
                val first = trimmedName.split(" ").firstOrNull { it.isNotEmpty() }
                if (!first.isNullOrEmpty()) return first
            }

            if (!email.isNullOrEmpty()) {
                val localPart = email.substringBefore("@")
                val namePart = localPart.split('.', '_', '-').firstOrNull { it.isNotEmpty() }
                if (!namePart.isNullOrEmpty()) {
                    return namePart.replaceFirstChar { if (it.isLowerCase()) it.titlecase() else it.toString() }
                }
            }

            return if (provider == AuthProvider.Guest) "Friend" else "User"
        }

    companion object {
        val guest = UserProfile(
            id = "guest_local",
            email = null,
            fullName = "Guest User",
            avatarUrl = null,
            provider = AuthProvider.Guest,
            createdAt = System.currentTimeMillis()
        )
    }
}

sealed interface AuthSessionState {
    data object Initializing : AuthSessionState
    data object Unauthenticated : AuthSessionState
    data class Authenticated(override val profile: UserProfile) : AuthSessionState
    data object Guest : AuthSessionState

    val isAuthenticatedOrGuest: Boolean
        get() = this is Authenticated || this is Guest

    val profile: UserProfile?
        get() = when (this) {
            is Authenticated -> profile
            is Guest -> UserProfile.guest
            else -> null
        }
}
